src/lib/util.ts

```diff
@@ -260,33 +271,45 @@ export function sleep(milliseconds: number): Promise<void> {
  * return network id, independent of the presence of Arc
  * @param web3Provider
  */
-export async function getNetworkId(web3Provider?: any): Promise<string> {
-  let arc: any;
-  let web3: any;
+export async function getNetworkId(web3Provider?: Web3Provider): Promise<string> {
+  let arc: Arc;
 
   try {
     arc = getArc();
   } catch (ex) {
     // Do nothing
   }
 
-  /**
-   * make sure that if the web3Provider is passed in, then the web3 we use matches it
-   */
-  if (arc && arc.web3 && (!web3Provider || (arc.web3.currentProvider === web3Provider))) {
-    web3 = arc.web3;
-  } else if ((window as any).web3 &&
-    (!web3Provider || ((window as any).web3.currentProvider === web3Provider))) {
-    web3 = (window as any).web3;
+  if (arc) {
+    const network = await arc.web3.getNetwork();
+    return network.chainId.toString();
   } else if (web3Provider) {
-    web3 = new Web3(web3Provider);
-  }
-
-  if (!web3) {
+    if (typeof web3Provider === "string") {
+      const provider = new JsonRpcProvider(web3Provider);
+      const network = await provider.getNetwork();
+      return network.chainId.toString();
+    } else if (Signer.isSigner(web3Provider)) {
```

Comment:
What cases does this `isSigner` logic cover?  Can we comment that in the code?

Reply:
The type `Web3Provider` could either be an RPC URL string, async sendable interface (web3.js provider), or an in-app signer: https://github.com/daostack/arc.js/blob/ad9d4c49dd4d25170a6df7e37d5c084f81cb57aa/src/utils/types.ts#L13

The signer case must be handled here because it could be any of these 3 types. The signer case is when you'd like to have a private key stored in the app's memory. More information can be found here https://github.com/daostack/arc.js/pull/450 & here https://github.com/daostack/arc.js/issues/445

Reply:
Can you please put this information in the code?